### ISSUE TITLE
Use predefined sorting for Grid data

### ISSUE THREAD

**Issue by pekam** [vaadin/vaadin-grid-flow#13]
@pleku commented on [Wed Nov 01 2017](https://github.com/vaadin/flow/issues/2818)

When showing a grid to the users, I want to show the data as sorted, so that they see the data in a meaningful order and can see that also from the grid that it is sorted this way.

*This feature depends on whether we can update the sort order in the `vaadin-grid-sorter`, meaning that it is exposing the ordering of the sorted columns in `multi-sort` mode.*

```java
// the following should throw, unless the column has been defined as sortable
column.sort(SortDirection.ASCENDING/DESCENDING/NONE);
```

This needs to update the sorting indicator in the UI e.g. `vaadin-grid-sorter direction="ascending"` and also update the sort orders on the server side and trigger new sorting of data. If multi-sort is enabled, then the appends to the sort order list.

If sorting is applied to the `DataProvider`, that will not be visible in the grid on the UI.

UPDATE: The sort-method was implemented in #290, but it works only after the grid is already rendered.

**Comment by schaerl** [comment]
Since the ticket is still open, I assume you consider this issue not yet fixed, and I just wanted to make sure if that is actually the case, or if I'm missing something: I'm using Vaadin 13alpha4, and using the sort API does not allow me to sort the grid before it's shown to the user. I.e. this does not work:

    public class MyView extends VerticalLayout{
        public MyView(){
            Grid grid = /* create grid */
            grid.setItems(itemList);
            grid.sort(/*List containing a sort order */);
        }
    }

The grid shown contains the unordered items (the same also happens when using a backend dataprovider).

If I set a breakpoint in the `fetch` method of ListDataProvider where it checks if a comparator is present, It will actually find the comparator and use it when first fetching the items. If I let the debugger run further, the UI is rendered, with the data sorted as I want it, but then the program runs into the breakpoint again, without any sort order. If I continue execution then, the grid items are rendered again but this time unordered

**Comment by pvnhome** [comment]
```
public class MyView extends VerticalLayout{
    public MyView(){
        Grid grid = /* create grid */
        grid.grid.setDataProvider(provider);
        grid.sort(/*List containing a sort order */); // not working (as described above)

        Button btn = new Button("Sort", event -> {
           grid.sort(/*List containing a sort order */); // working as expected
        });
    }
}
```

**Comment by pvnhome** [comment]
```
public class MyView extends VerticalLayout{
    public MyView(){
        Grid grid = /* create grid */
        grid.grid.setDataProvider(provider);

        // working but not visible in the grid
        provider.setSortOrders(/*List containing a sort order */);
    }
}

**Comment by pvnhome** [comment]
Clone and run:

```
git clone https://github.com/pvnhome/grid_sorting_example.git
cd grid_sorting_example
gradlew appRun
```

Project should soon then be running on http://localhost:8080/sortgrid/

You can see described above behavior in simplified log:

```
SortgridView
sizeInBackEnd: 200, sort=[], filter=absent
fetchFromBackEnd: offset=0, limit=50, sort=[id (D)], filter=absent
sizeInBackEnd: 200, sort=[], filter=absent
fetchFromBackEnd: offset=0, limit=50, sort=[], filter=absent
fetchFromBackEnd: offset=50, limit=100, sort=[], filter=absent
```

On button click sorting working as expected:


```
button: sort 1
sizeInBackEnd: 200, sort=[], filter=absent
fetchFromBackEnd: offset=0, limit=150, sort=[c (A),id (A)], filter=absent

button: filter on
sizeInBackEnd: 100, sort=[], filter=[need to filter]
fetchFromBackEnd: offset=0, limit=100, sort=[c (A),id (A)], filter=[need to filter]
```

## No sorting


Variant without initial sorting:

```
git checkout without_initial_sort
gradlew appRun
```

Log:


```
SortgridView
sizeInBackEnd: 200, sort=[], filter=absent
fetchFromBackEnd: offset=0, limit=50, sort=[], filter=absent
fetchFromBackEnd: offset=50, limit=100, sort=[], filter=absent
```

## Sorting using provider


```
git checkout provider_sort
gradlew appRun
```

Demonstrated code:

```java
public class MyView extends VerticalLayout{
    public MyView(){
        Grid grid = /* create grid */
        grid.grid.setDataProvider(provider);

        // working but not visible in the grid
        provider.setSortOrders(/*List containing a sort order */);
    }
}
```

Log:


```
SortgridView
sizeInBackEnd: 200, sort=[id (D)], filter=absent
fetchFromBackEnd: offset=0, limit=50, sort=[id (D)], filter=absent
fetchFromBackEnd: offset=50, limit=100, sort=[id (D)], filter=absent
```

Sorting working but not visible in the grid.

## Without filtering and delay

Variant without filtering and delay:


```
git checkout no_filter_and_delay
gradlew appRun
```

**Comment by estronque** [comment]
Hi.
The same wrong behavior still happens in vaadin version 13.0.6.
I have found an ugly workaround by calling the sort method asynchronously 1 second after entering the view using push functionality, but the data records changes are very noticeable.

**Comment by pvnhome** [comment]
Vaadin version in grid_sorting_example updated to 13.0.6

**Comment by heruan** [comment]
Sorting still doesn't work in 4.0.0.rc3 if set before the first render; I really hope this will be fixed before 4.0.0 is released! Any update on this @pekam?

**Comment by pekam** [comment]
I don't think it's planned to be fixed for V14, as the product is owned by the Flow team who are busy with finalizing the framework. Also, there are other prioritized bugs for Grid waiting to be fixed at the moment.

For the reference, if someone starts working on this, I took a look at the issue. The problem is that when the web component is initialized, it notifies [this observer in gridConnector.js](https://github.com/vaadin/vaadin-grid-flow/blob/6e80b784a9d301045bf46f59ee7a7b510daa8ed0/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js#L406) with empty sorters, notifying the server to reset the sort orders. It might also be that [setting the sort orders from server to client](https://github.com/vaadin/vaadin-grid-flow/blob/6e80b784a9d301045bf46f59ee7a7b510daa8ed0/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js#L415) is done too soon, and the web component overrides those values with the initial empty sort orders.

I might try to find time to fix it myself during next week, as I've already come close to the solution (hopefully). Not making promises though.
